### PR TITLE
[ New Test ] (250315@main): [ macOS Debug ] TestWTF.WTF_SuspendableWorkQueue.Suspend is a flaky failure

### PR DESCRIPTION
**Pull request by szewai**
#### a8ab43132b0d66e1dd1fc9d519d22710fa897962
<pre>
[ New Test ] (250315@main): [ macOS Debug ] TestWTF.WTF_SuspendableWorkQueue.Suspend is a flaky failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=245793">https://bugs.webkit.org/show_bug.cgi?id=245793</a>
&lt;rdar://100517874&gt;

Reviewed by Chris Dumez.

We used to unset suspended flag when task runs, but all tasks were submitted before suspend(), so they may complete
before suspension and suspended flag was not updated. To ensure the flag is updated, now we schedule some tasks after
suspend().

* Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/255161@main">https://commits.webkit.org/255161@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2c85becc060757fe05752e614652d54fdd742550

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/694 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101254 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161305 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95544 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/699 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83870 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97605 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/436 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27389 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82360 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70429 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35666 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33438 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17141 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3585 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37258 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36258 "Passed tests") | |

### CODE REVIEW

File: Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp [hunk 1]
```diff
@@ -37,19 +38,19 @@ TEST(WTF_SuspendableWorkQueue, Suspend)
...
     bool suspended = false;
+    // Main thread only.
     bool suspendCompletionHandlerCalled = false;
+    bool taskAfterResumeCompleted = false;
+
     auto queue = SuspendableWorkQueue::create("com.apple.WebKit.Test.simple");
     for (int i = 0; i < taskCount; i ++) {
         queue->dispatch([&]() mutable {
             Locker locker { lock };
-            suspended = false;
```
Comment (cdumez) [inline Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp:47]: If I understand correctly from your commit log, what would happen is that all tasks could finish running before the call to suspend() (or at least before the request to suspend is processed).

As a result, you are now updating suspended and allTasksCompleted in new tasks that are scheduled after the queue is resumed.

However, wouldn't it mean that we would still fail the following check below?
`EXPECT_LE(completedTaskCount, taskCount);`

Maybe we should:
1. Schedule some tasks
2. suspend
3. Schedule some more tasks
4. Wait until suspension is complete.
5. Make sure not all tasks were completed
6. Sleep a bit and make sure no more tasks complete
7. Resume the work queue
8. Make sure all tasks complete.

Comment (szewai) [reply Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp:47]: Why would the check fail? This is `EXPECT_LE` not `EXPECT_LT`.
But the test flow sounds reasonable; I will make the change.

Comment (cdumez) [reply Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp:47]: You're right, I misread the EXPECT check.

File: Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp [hunk 2]
```diff
@@ -35,36 +36,64 @@ namespace TestWebKitAPI {
...
     {
         Locker locker { lock };
-        EXPECT_LE(completedTaskCount, taskCount);
+        EXPECT_LT(completedTaskCount, taskCount);
         EXPECT_TRUE(suspended);
+        lastCompletedTaskCount = completedTaskCount;
+    }
+
+    // Ensure no more task is processed.
+    Util::runFor(0.1_s);
```
Comment (cdumez) [inline Tools/TestWebKitAPI/Tests/WTF/SuspendableWorkQueue.cpp]: nit: 0.1_s ->. 100_ms ?

**Review by cdumez — APPROVED** [review]
LGTM

**Pull request merged**